### ISSUE TITLE
Repeat.for to support key (custom resolution strategy)

### ISSUE THREAD

**Issue by tomasbonco** [aurelia/aurelia#1264]
# 🙋 Feature Request

While React and Vue do support additional `key` attribute to signalize strategy, which elements to create, Aurelia and Aurelia 2 (as far as I can tell from docs) relies only on indexes and array manipulation. This is an issue when the whole array is replaced.

## 🤔 Expected Behavior

```html
<div repeat.for="todo of todos & key: 'id'"> ${ todo.text } </div>
<!-- or --> 
<div repeat.for="todo of todos" key.bind="todo.id"> ${ todo.text } </div>
<!-- or similar -->
```

```js
let todos = [ { text: 'a', id: 1 }, { text: 'b', id: 2 } ];
// render
todos = [ { text: 'a', id: 1 }, { text: 'c', id: 3 }, { text: 'b', id: 2 } ];

// Produces:

// INITIAL STATE (HTML elements)
a
b

// NEW STATE (HTML elements)
a - reused
c - newly created element
b - reused
```
Every created HTML element gets an id. This id must be unique for each element in the array. When `repeat.for` renders the array again (due to a change), it searches for an existing HTML element with the given id first, and reuses it. When such HTML element cannot be found a new element is created.

## 😯 Current Behavior

```js
/* ONLY WAY HOW TO CURRENTLY ACHIEVE THE EXPECTED RESULT */

const array = [ 'a', 'b' ];
// render
array.slice( 1, 'c' );

// Produces:

// INITIAL STATE (HTML elements)
a
b

// NEW STATE (HTML elements)
a - reused
c - newly created element
b - reused
```

```js
/* CURRENT BEHAVIOR IN PROBLEMATIC SITUATION */

let array = [ 'a', 'b' ];
// render
array = [ 'a', 'c', 'b']

// Produces:

// INITIAL STATE (HTML elements)
a
b

// NEW STATE (HTML elements)
a - reused
c - reused
b - newly created element
```

## 💁 Possible Solution

Vue: https://vuejs.org/v2/guide/list.html
React: https://reactjs.org/docs/lists-and-keys.html#keys

## 🔦 Context

I fetch data periodically from a server. I have to animate newly added items (and also removed). Currently, I have to compare the new array with the existing array and "simulate" changes. Not only is that difficult, but doesn't work well on unsorted array, or in a situation when the value for the sorting key changes for a given element (for example speed has changed but the driver is known, thus the element should be way lower, but shouldn't animate).

I would like to simply replace the array.

**Comment by bigopon** [comment]
@tomasbonco thanks for this issue, the current behavior is undesirable, and should be replaced with a better way of re-rendering on update/mutation. It may work similarly like v1 `matcher`, or `key` based like other frameworks.

**Comment by tomasbonco** [comment]
Hi, thank sou for solving this for v2. I have additional one question about v1 - How does `matcher` solves the situation for v1?
I can't get my head around how it works. I have a list of items, this list of items is from backend and is refreshed often. Using just `repeat.for` causes that whole list is rerendered. Not only is that expensive, it also breaks the animations. Here is a code:
```html
<div repeat.for="item of items">
    ...
</div>
```

I hoped I could use:
```html
<div repeat.for="item of items" matcher.bind="compareIds">
   ...
```

**Comment by bigopon** [comment]
did you try `matcher.bind="compareIds"`?
How does the `compareIds` look like in your view model? Also, is this issue about v1 or v2?

**Comment by tomasbonco** [comment]
1. Yes, this is v1 issue.
2. I tried `matcher.bind="compareIds"`
3. I tried `Repeat.useInnerMatcher = false;`

The only way it seems to work is when `repeat.for` and `matcher` are not at the same element (I was now trying all combinations). In my opinion that's weird, because I expect that `matcher` is somehow connected to the element it is applied to. However I would like whole content of the `<template>` element to be handled the same, not just one child.

This is the only solution, that works for me:
```html
<template repeat.for="item of items">
    <div matcher.bind="compareIds"> ... </div>
</template>
```

compareIds is pretty straight-forward:
```js
compareIds( a, b )
{
	console.log( 'called', a, b )
	return a.id === b.id;
}
```
